### DISCUSSION TITLE
Unity Integration

### DISCUSSION THREAD

**Discussion by djynnantonic** [deepgram/community#163] [General help] [answered]
I'm having difficulty with integrating Unity into Deepgram.  I got it working in native c# after just a few hours which is encouraging.  

When using Deepgram's sample Unity project, it throws an "Unable to connect to the remote server" at websocket.OnError, using the recommended Unity 2020.3.30.

I've tried multiple ways to bring the Deepgram library into Unity but there are many dependent Microsoft and System libraries.  I can get the build to the point where it recognizes all of them with no dependency errors, and Deepgram is able to run the creation of credentials but will freeze/not return from the initial call to deepgramClient = new DeepgramClient(credentials);.

Has anyone completed a successful Unity integration, and if so what were the conditions?  Thanks!

**Comment by nikolawhallon** [comment]
It sounds like you tried this but it was not working for you? I’ve just tested it on Mac and Windows (but not yet Linux): 

https://blog.deepgram.com/deepgram-unity-tutorial/

we also have this which is tested and working on Linux, Windows, and Quest 2:

https://github.com/deepgram-devs/DGverse

both of these projects use the Native WebSocket library in order to connect to Deepgram:

https://github.com/endel/NativeWebSocket

Note that the Deepgram .NET SDK won’t work with Unity (without possibly major headaches), but hopefully these two projects can help. If not, can you share some of your code for the Deepgram part and I can take a look?

**Reply by djynnantonic** [reply]
Hi Nikola - I'm still having issues with the git project and with a careful line by line recreation.  I'm using Unity 2020.3.30, Windows 11, the WebSocket plugin installed by git, and checked that my key works with my c# standalone implementation. I'm not so interested in the ball so I removed that portion of the code.  I'm out of ideas and wondering what else it might be. Here's what I have: 

using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using NativeWebSocket;
using System.Text.RegularExpressions;

[System.Serializable]
public class DeepgramResponse
{
    public int[] channel_index;
    public bool is_final;
    public Channel channel;
}

[System.Serializable]
public class Channel
{
    public Alternative[] alternatives;
}

[System.Serializable]
public class Alternative
{
    public string transcript;
}

public class DGSupport : MonoBehaviour
{
    WebSocket websocket;

    // Start is called before the first frame update
    async void Start()
    {
        var headers = new Dictionary<string, string>
            {
                { "Authorization", "[my key which works in c# standalone implementation]" }
            };
        websocket = new WebSocket("wss://api.deepgram.com/v1/listen?encoding=linear16&sample_rate=" + AudioSettings.outputSampleRate.ToString(), headers);

        Debug.Log("Audio settings " + AudioSettings.outputSampleRate.ToString());  // reads as 48k

        websocket.OnOpen += () =>
        {
            Debug.Log("Connected to Deepgram!");
        };

        websocket.OnError += (e) =>
        {
            Debug.Log("Error: " + e);
        };

        websocket.OnClose += (e) =>
        {
            Debug.Log("Connection closed!");
        };

        websocket.OnMessage += (bytes) =>
        {
        };

            await websocket.Connect();

    }


    // Update is called once per frame
    void Update()
    {
#if !UNITY_WEBGL || UNITY_EDITOR
        websocket.DispatchMessageQueue();
#endif
    }

    private async void OnApplicationQuit()
    {
        await websocket.Close();
    }

    public async void ProcessAudio(byte[] audio)
    {
        if (websocket.State == WebSocketState.Open)
        {
            await websocket.Send(audio);
        }
    }

}

I've traced it in WebSocket.cs from the websocket.Connect() and it fails at line 471 "await m_Socket.ConnectAsync(uri, m_CancellationToken);", going right to the catch and not hitting the "OnOpen?.Invoke();" in the next line.  No compiler errors, and all attempts end with "Unable to connect to the remote server".  Same results with the second sample project using Cleverbot, and I also tried with my firewalls off.

Something in the environment perhaps?  Is there a port I should enable or something along those lines? Any pointers would be greatly appreciated, thanks!

**Reply by nikolawhallon** [reply]
Let me try with Unity 2020.3.30 and Windows 11! Does it both not work in the Editor nor in an export of the project (Windows export or HTML/WebGL export)? Might help narrow down the issue.

**Reply by djynnantonic** [reply]
Aha!  I didn't try it as a build.  Let me try that and let you know.

**Reply by nikolawhallon** [reply]
(either way, it ought to be working in the Editor too, so I'll try to get a reproducer in Windows 11 and figure this out)

**Reply by djynnantonic** [reply]
Unfortunately, compiling and running the Unity version yields the same results.  Still perplexed as the .NET c# standalone implementation works, meaning the connection to the servers is working from my computer.  Could it be the embedded init string? It references v1 and in using the .NET version I thought it references v2.

**Reply by djynnantonic** [reply]
(Also tried Android in the editor and on device with same results.)

**Reply by nikolawhallon** [reply]
I tried with VirtualBox - no luck, my Windows 11 VM would crash when I tried Unity, but then I provisioned a machine with Windows 11 (`Win11_22H2_English_x64v2.iso`), Unity 2020.3.30, Git for Windows, and vscode, and the demo/repo worked for me without modification (except to insert my Deepgram API key).

The Deepgram ASR API endpoint should be:
```
wss://api.deepgram.com/v1/listen
```
In the .NET SDK I see the reference to this here:
```
Deepgram/Transcription/LiveTranscriptionClient.cs:                var wssUri = GetWSSUriWithQuerystring("/v1/listen", options);
Deepgram/Transcription/PrerecordedTranscriptionClient.cs:                "/v1/listen",
Deepgram/Transcription/PrerecordedTranscriptionClient.cs:                "/v1/listen",
```
there may have been at one point references to a `v2` for an earlier API version (yes, `v2` -> `v1` is very interesting!)

I will dig a bit more, but is there anything else about your environment that might be relevant? I assume it's not a VM and just a Windows 11-installed laptop/desktop.

**Reply by djynnantonic** [reply]
Good to know it is working out of the box in the default environment for you.  

When I was trying to get compatible .NET DLLs together to run in Unity, I vaguely recall the Deepgram DLL being V2, but that is of course different than the network message contents.  Thanks for confirming that too. 

I am using a Dell gamer PC, no virtualization (though now that I think of it, I have enabled some Windows virtualization extensions I can try to disable).  I have a second Windows 11 gamer laptop I will install on and see if I get any different results.

Thanks for the thorough investigation!

**Reply by djynnantonic** [reply]
Additional info: I re-downloaded the demo Unity project on a second Windows computer (Windows 11, Unity 2020.3.30), compiled it in Unity there, and it doesn't connect to the server either.  It does connect from the second computer using the c# standalone version.  It seems like it is distributed with NativeWebSocket but just in case I uninstalled it and reinstalled from github, no difference.

The virtualization extension I was thinking of is called Hyper-V but was not enabled on either computer.

How about sending me a Windows build created in your environment and let me run it here?  If it works, we will have isolated it to the build environment; if it doesn't, it will suggest something at the OS level?

**Reply by nikolawhallon** [reply]
This is fascinating! Though I'm sorry it's happening. Attached is my build - the API key associated with it will expire so it won't work forever. I had to make this change to my code for it to work (and I might as well update the Unity blog post too - I think I used what I did for the JSON at the time because of a quirk with the Unity version on Mac):

```
            //DeepgramResponse deepgramResponse = new DeepgramResponse();
            //object boxedDeepgramResponse = deepgramResponse;
            //EditorJsonUtility.FromJsonOverwrite(message, boxedDeepgramResponse);
            //deepgramResponse = (DeepgramResponse) boxedDeepgramResponse;
            DeepgramResponse deepgramResponse = JsonUtility.FromJson<DeepgramResponse>(message);
```
The attached build works correctly for me on my Windows 11 laptop, but I don't have another machine to test with. It's probably worth opening a GitHub issue on/in the NativeWebSocket repo, there have been a handful of similar issues from users, but they have all had solutions which indicate something different.

[UnityDeepgramDemoBuild.zip](https://github.com/deepgram/community/files/11710257/UnityDeepgramDemoBuild.zip)

**Reply by nikolawhallon** [reply]
(it's annoyingly full-screen, Alt + F4 will quit it though)

**Reply by djynnantonic** [reply]
Thanks.  Update: your build works on both my desktop and laptop computers, which implies it's something in the build environment.  It couldn't be my router settings, firewall settings, or anything like that, because yours works (as does my c# .NET).

I'd like to rule out the input devices because I never see the error from MicrophoneObject.Start().  You will know better than me though if they could be related, but again, no error messages.  Which points to a network-related build issue.

I've removed NativeWebSockets (which causes errors) and reinstalled it (which fixes them) with the Package Manager a couple times.  The source archive comes with it already installed in the project.

Are Project Settings set in the project archive, could there be a difference in our settings?  Can you think of any other tests?  I think if I send you a build it won't work but glad to do so if it will shed light on the issue.

**Reply by djynnantonic** [reply]
I'm also thinking that it could be something simple, trivial, easy that has been overlooked... a real "d'oh" moment.  We'll never find a bug in the code that isn't there if it's caused something else.

**Reply by djynnantonic** [reply]
If you have a chance, here's my source code, it will need a valid key.  Whether or not this runs in your environment could help triangulate the issue.
[UnityDeepgramDemo-main.zip](https://github.com/deepgram/community/files/11710740/UnityDeepgramDemo-main.zip)

**Reply by nikolawhallon** [reply]
I tried your project and got the errors you got! I can't seem to attach very much, I am hitting the 25 MB limit. But what I did was I tried replacing parts of your project with parts of mine, not very thoughtful or anything, but can you try replacing your "Assets" directory with this one:
[Assets.zip](https://github.com/deepgram/community/files/11711416/Assets.zip)
I haven't been able to look harder at the differences between that/my one and yours, but I found that when I replaced yours with mine, your project then worked. I am pretty slow navigating in Windows, but now I'm eager to try out your project elsewhere and more to narrow down the issue, but do give the attached Assets directory a try.

**Reply by djynnantonic** [reply]
Thanks for continuing to look at the issue.  Unfortunately that asset folder (with my key) still fails in the same way.  

I'm relieved that you were able to duplicate the problem with my source, but confused that your solution didn't fix it on my end.  The only thing that would be different is the key, and I've triple checked that my key is the same one that works in .NET .

Perhaps you could zip your entire working project directory?  (You can delete 'library' which will shrink the zip & regenerate here.)  Then the only difference would be the key, or the Unity libraries installed in 2020.3.30, but I did a fresh Unity install for this work.

Thanks again for your tenacity.  In the words of Lewis Carroll, "Curiouser and Curiouser "!

**Reply by nikolawhallon** [reply]
Here is the link to the full project where I replaced the Assets directory (and it's working for me): https://drive.google.com/file/d/14gtSb4C9m4iJ7i3e22u4oDxzZ5yJD--p/view?usp=sharing

And attached is the same thing, minus the Library directory - somehow that directory may be key! I'm certain the Deepgram API key can't be to blame (at worst would result in a 403 or so). If these two zips don't help solve the mystery, I could pair sometime, I am very interested in seeing what is going on!

[UnityDeepgramDemo-main-no-library.zip](https://github.com/deepgram/community/files/11728860/UnityDeepgramDemo-main-no-library.zip)

**Reply by djynnantonic** [reply]
Thanks for the continued efforts, but both fail in the same way :(  What development environment are you using?  I'm going to install Unity 2020.3.30 on my Mac and try there.

**Reply by nikolawhallon** [reply]
This is on a thinkpad with Windows 11 freshly installed, Unity 2020.3.30, vscode, and Git for Windows also installed (I did not even install the .NET SDK).

**Reply by djynnantonic** [reply]
Wow.  Now I'm thinking it's something super simple.  I'll let you know how the Mac goes tomorrow.  I'm relieved that you saw it fail :D  Thanks again.

**Reply by nikolawhallon** [reply]
![image](https://github.com/deepgram/community/assets/43757888/a16fb5bb-2e16-477a-963c-3eaf2d4bcf78)

on the left is your code verbatim (which indeed does not work for me, in the same way you describe!), on the right is your project with the Assets directory replaced - this is beyond my knowledge right now, but something is different about the ProjectCache

EDIT: this might just be something as boring as the file paths being different..

**Reply by djynnantonic** [reply]
Hi Nikola - Still no positive results. I've downloaded each example from this thread with a clear head and confirmed the results.

Stepping way back, what we have is:

* The Windows Unity .EXE build you sent me works on my two Windows machines
* I can build and connect to the servers in .NET c#
* None of the attempts to build from your or my Unity sources works on either of my machines, including the one that you compile that works for me
* You were able to duplicate my Unity build failures and fix them with a new Assets folder, which doesn't work for me

Other simple things I tried here were selecting different microphones, using a newer Unity (2021.3.12), power cycling my router, trying configurations with my firewall off, doing a Git pull of the project instead of a .ZIP (I have another project where that was a 'thing').

The Mac Unity build didn't work in the editor for me either, but developing there is slow going and I didn't try as much as I could.

It boils down to: if you build it, it works on my machines; if I build it, it doesn't.  I'm still puzzling through exactly what that means and what possible solutions are, but it does narrow it down somewhat.

**Reply by djynnantonic** [reply, marked as answer]
I had a colleague look at the materials and he solved it.  We suspected it would either be an obscure technical detail or a Homer Simpson moment.  It turns out the "key" field requires the string "token" prior to the actual key (which he missed on the first go too).  We never detected that in each other's sources because we of course removed our keys.  In conclusion, "D'oh!"

Thanks for the diligent work and I'm up and running!

**Reply by nikolawhallon** [reply]
OMG wow, no wonder no amount of passing back and forth the code worked because, like you said, that's the one piece of code we self-censored ; ; so happy this wasn't some extremely obscure thing - but also a bit embarrassed I didn't notice O_O
